### PR TITLE
WEEK 6

### PR DESCRIPTION
**Pull request by spartak0**

### CODE REVIEW

File: app/src/main/java/com/example/weather/data/db/converter/CurrentConverter.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.example.weather.data.db.converter;
+
+import androidx.room.TypeConverter;
+
+import com.example.weather.data.db.entity.forecast_detail.Current;
+import com.example.weather.data.db.entity.forecast_detail.Weather;
+
+public class CurrentConverter {
+    @TypeConverter
+    public float fromHobbies(Current current) {
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/data/db/converter/CurrentConverter.java:10]: Where is hobby?

File: app/src/main/java/com/example/weather/data/db/entity/forecast_detail/Current.java [hunk 2]
```diff
@@ -4,6 +4,11 @@
     float temp;
     Weather[] weather;
 
+    public Current(float temp, Weather[] weather) {
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/data/db/entity/forecast_detail/Current.java:7]: What is `Current`? Better to use `CurrentForecast` for example

File: app/src/main/java/com/example/weather/domain/Repository.java [hunk 3]
```diff
@@ -28,8 +30,7 @@
     Completable updateWeather(WeatherData weatherData);
 
     Observable<Pair<Float,String>> getCurrentWeatherDataByCoord(String lat, String lon, String units);
-
-    Observable<HashMap<String,String>> getDailyWeatherDataByCoord(String lat, String lon,int day , String units);
+    Observable<Pair<Float,String>> getDailyWeatherDataByCoord(String lat, String lon, int day , String units);
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/domain/Repository.java]: You pass `units` to every request like argument. Could we set it once for every request using `Interceptor` for example?

File: app/src/main/java/com/example/weather/domain/mapper/HourlyMapper.java [hunk 4]
```diff
@@ -1,26 +1,53 @@
...
+        this.context=context;
+    }
 
+    @RequiresApi(api = Build.VERSION_CODES.N)
     @Override
     public ArrayList<Triple<String, String, String>> toDomain(WeatherEntity weatherEntity) {
         ArrayList<Triple<String, String, String>> list = new ArrayList<Triple<String,String,String>>();
         Hourly[] hourlies = weatherEntity.getHourly();
+        SharedPreferences preferences = this.context.getSharedPreferences("Settings", MODE_PRIVATE);
+        String language = preferences.getString("MyLang","");
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/domain/mapper/HourlyMapper.java]: It's glad to see usage of `SharedPrefs` but also I'll recommend to use constants for "MyLang" and "Settings".

File: app/src/main/java/com/example/weather/ui/forecasts/all_forecast/AllForecastFragment.java [hunk 5]
```diff
@@ -30,14 +33,16 @@
...
+    ForecastItemAdapter adapter;
+    SettingManager settingManager;
 
     @Nullable
     @Override
     public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
         binding = FragmentForecastBinding.inflate(inflater, container, false);
         View view = binding.getRoot();
         getActivity().setTitle(R.string.location);
+        settingManager= new SettingManager(getContext().getApplicationContext());
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/forecasts/all_forecast/AllForecastFragment.java]: It should be inside `Repository` and be able to use via `ViewModel`

File: app/src/main/java/com/example/weather/ui/main/MainActivity.java [hunk 6]
```diff
@@ -26,6 +30,19 @@ protected void onCreate(Bundle savedInstanceState) {
         settingManager= new SettingManager(getBaseContext());
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/main/MainActivity.java]: See upper comment

File: app/src/main/java/com/example/weather/ui/main/MainActivity.java [hunk 7]
```diff
@@ -48,7 +65,7 @@ public boolean onOptionsItemSelected(@NonNull MenuItem item) {
                 recreate();
                 break;
             case R.id.english:
-                 settingManager.setLocale("en");
+                settingManager.setLocale("en");
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/main/MainActivity.java]: Better to use constants for "en" string and others

File: app/src/main/java/com/example/weather/utils/SettingManager.java [hunk 8]
```diff
@@ -26,9 +28,32 @@ public void setLocale(String lang){
         editor.putString("MyLang", lang);
         editor.apply();
     }
-    public void loadLocale(){
+    public void  loadLocale(){
         SharedPreferences preferences = context.getSharedPreferences("Settings", MODE_PRIVATE);
         String language = preferences.getString("MyLang","");
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/utils/SettingManager.java]: See upper comments

**Review by YellowShark — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/weather/data/db/entity/forecast_detail/Current.java (hunk 2 above)
Comment (spartak0) [reply app/src/main/java/com/example/weather/data/db/entity/forecast_detail/Current.java:7]: Поля должны называть соответствующе как и приходят с запроса

File: app/src/main/java/com/example/weather/ui/main/MainActivity.java [hunk 9]
```diff
@@ -11,23 +12,22 @@
...
 import com.example.weather.databinding.ActivityMainBinding;
+import com.example.weather.utils.Constant;
 import com.example.weather.utils.SettingManager;
 import com.google.android.material.dialog.MaterialAlertDialogBuilder;
 
 public class MainActivity extends AppCompatActivity {
 
     ActivityMainBinding binding;
-    SettingManager settingManager;
+    SettingManager settingManager= RepositoryImpl.getInstance().getSettingsMenager();
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/main/MainActivity.java]: No, you didn't catch. It should be only like a method inside `ViewModel`. So, SettingManager's instance inside repository and repository use methods (`setLocale(...)`), `ViewModel` has instance of the `Repository` and invokes `setLocale(...)` and then you can invoke this method via `ViewModel`. This logic is similar with usage of network, db and other services. We shouldn't depend on realisation.

File: app/src/main/java/com/example/weather/ui/main/MainActivity.java [hunk 10]
```diff
@@ -61,11 +61,11 @@ public boolean onCreateOptionsMenu(Menu menu) {
     public boolean onOptionsItemSelected(@NonNull MenuItem item) {
         switch (item.getItemId()) {
             case R.id.russian:
-                settingManager.setLocale("ru");
+                settingManager.setLocale(Constant.RU);
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/main/MainActivity.java]: I hope it will transform into `viewModel.setLocale(...)`

**Review by YellowShark — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/weather/data/RepositoryImpl.java [hunk 11]
```diff
@@ -109,8 +109,29 @@ public Completable updateWeather(WeatherData weatherData) {
...
-        return this.settingManager;
+    public Observable<String> getTimezone(String lat, String lon) {
+        return ForecastApi.Instance.getForecastApi().getWeatherDataByCoord(lat,lon)
+                .map(WeatherEntity::getTimezone);
+    }
+
+
+    @Override
+    public void loadLocale(Context context) {
+        if(settingManager==null) this.settingManager=new SettingManager(context);
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/data/RepositoryImpl.java]: Why do you do that? As I see, Repository is Singleton. Moreover, when you init Repository you pass Context, so you don't need to pass Context here. You can just init SettingManager with Repository.

**Review by YellowShark — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/weather/data/RepositoryImpl.java (hunk 11 above)
Comment (spartak0) [reply app/src/main/java/com/example/weather/data/RepositoryImpl.java]: To change the language, I need the context from the main activity. The repository is initialized in the app.

**Review by YellowShark — APPROVED** [review]

**Pull request merged**